src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,732 @@
+extern "C" IMAGE_DOS_HEADER __ImageBase;
+
+const UINT WM_TASKBARCREATED = RegisterWindowMessage(L"TaskbarCreated");
+
+WindowEmperor::WindowEmperor() noexcept :
+    _app{}
+{
+    _manager.FindTargetWindowRequested([this](const winrt::Windows::Foundation::IInspectable& /*sender*/,
+                                              const winrt::Microsoft::Terminal::Remoting::FindTargetWindowArgs& findWindowArgs) {
+        {
+            const auto targetWindow = _app.Logic().FindTargetWindow(findWindowArgs.Args().Commandline());
+            findWindowArgs.ResultTargetWindow(targetWindow.WindowId());
+            findWindowArgs.ResultTargetWindowName(targetWindow.WindowName());
+        }
+    });
+
+    _dispatcher = winrt::Windows::System::DispatcherQueue::GetForCurrentThread();
+}
+
+WindowEmperor::~WindowEmperor()
+{
+    _app.Close();
+    _app = nullptr;
+}
+
+void _buildArgsFromCommandline(std::vector<winrt::hstring>& args)
+{
+    if (auto commandline{ GetCommandLineW() })
+    {
+        auto argc = 0;
+
+        // Get the argv, and turn them into a hstring array to pass to the app.
+        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
+        if (argv)
+        {
+            for (auto& elem : wil::make_range(argv.get(), argc))
+            {
+                args.emplace_back(elem);
+            }
+        }
+    }
+    if (args.empty())
+    {
+        args.emplace_back(L"wt.exe");
+    }
+}
+
+bool WindowEmperor::HandleCommandlineArgs()
+{
+    std::vector<winrt::hstring> args;
+    _buildArgsFromCommandline(args);
+    auto cwd{ wil::GetCurrentDirectoryW<std::wstring>() };
+
+    Remoting::CommandlineArgs eventArgs{ { args }, { cwd } };
+
+    const auto isolatedMode{ _app.Logic().IsolatedMode() };
+
+    const auto result = _manager.ProposeCommandline(eventArgs, isolatedMode);
+
+    if (result.ShouldCreateWindow())
+    {
+        CreateNewWindowThread(Remoting::WindowRequestedArgs{ result, eventArgs });
+
+        _manager.RequestNewWindow([this](auto&&, const Remoting::WindowRequestedArgs& args) {
+            CreateNewWindowThread(args);
+        });
+
+        _becomeMonarch();
+    }
+    else
+    {
+        const auto res = _app.Logic().GetParseCommandlineMessage(eventArgs.Commandline());
+        if (!res.Message.empty())
+        {
+            AppHost::s_DisplayMessageBox(res);
+            ExitThread(res.ExitCode);
+        }
+    }
+
+    return result.ShouldCreateWindow();
+}
+
+void WindowEmperor::WaitForWindows()
+{
+    MSG message;
+    while (GetMessage(&message, nullptr, 0, 0))
+    {
+        TranslateMessage(&message);
+        DispatchMessage(&message);
+    }
+}
+
+void WindowEmperor::CreateNewWindowThread(Remoting::WindowRequestedArgs args)
+{
+    Remoting::Peasant peasant{ _manager.CreatePeasant(args) };
+
+    auto window{ std::make_shared<WindowThread>(_app.Logic(), args, _manager, peasant) };

```

Comment:
WindowThread could be `enable_shared_from_this` to reduce the likelihood that it gets constructed wrong